RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc

```diff
@@ -0,0 +1,276 @@
+  
+  if ( nrows > theBuffer.rows() || 
+       ncols > theBuffer.columns() ) 
+    { // change only when a larger is needed
+      // Resize the buffer
+      theBuffer.setSize(nrows,ncols);  // Modify
+      bufferAlreadySet = true;
+    }
+  
+  return true;   
+}
+//----------------------------------------------------------------------------
+//!  \brief Cluster hits.
+//!  This method operates on a matrix of hits
+//!  and finds the largest contiguous cluster around
+//!  each seed hit.
+//!  Input and output data stored in DetSet
+//----------------------------------------------------------------------------
+void MTDThresholdClusterizer::clusterize( const FTLRecHitCollection & input,
+					  const MTDGeometry* geom,
+					  const MTDTopology* topo,
+					  FTLClusterCollection& output) {
+  
+  FTLRecHitCollection::const_iterator begin = input.begin();
+  FTLRecHitCollection::const_iterator end   = input.end();
+  
+  // Do not bother for empty detectors
+  if (begin == end) 
+    {
+      edm::LogInfo("MTDThresholdClusterizer") << "No hits to clusterize";
+      return;
+    }
+
+  DEBUG("Input collection " << input.size());
+  assert(output.empty());
+
+  std::set<unsigned> geoIds; 
+  std::multimap<unsigned, unsigned> geoIdToIdx;
+  
+  unsigned index = 0;
+  for(const auto& hit : input) 
+    {
+      MTDDetId mtdId=MTDDetId(hit.detid());
+      if (mtdId.subDetector() != MTDDetId::FastTime)
+	{
+	  throw cms::Exception("MTDThresholdClusterizer") << "MTDDetId: " << std::hex
+							  << mtdId.rawId()
+							  << " is invalid!" << std::dec
+							  << std::endl;
+	}
+
+      if ( mtdId.mtdSubDetector() == MTDDetId::BTL )
+	{
+	  BTLDetId hitId(hit.detid());
+	  DetId geoId = hitId.geographicalId( (BTLDetId::CrysLayout) topo->getMTDTopologyMode() ); //for BTL topology gives different layout id
+	  geoIdToIdx.emplace(geoId,index);
+	  geoIds.emplace(geoId);
+	  ++index;
+	}
+      else if ( mtdId.mtdSubDetector() == MTDDetId::ETL )
+	{
+	  ETLDetId hitId(hit.detid());
+	  DetId geoId = hitId.geographicalId();
+	  geoIdToIdx.emplace(geoId,index);
+	  geoIds.emplace(geoId);
+	  ++index;
+	}
+      else
+	{
+	  throw cms::Exception("MTDThresholdClusterizer") << "MTDDetId: " << std::hex
+							  << mtdId.rawId()
+							  << " is invalid!" << std::dec
+							  << std::endl;
+	}
+    }
+
+  //cluster hits within geoIds (modules)
+  for(unsigned id : geoIds) {
+    //  Set up the clusterization on this DetId.
+    if ( !setup(geom,topo,DetId(id)) ) 
+      return;
+    
+    auto range = geoIdToIdx.equal_range(id);
+    DEBUG("Matching Ids for " << std::hex << id << std::dec << " [" <<  range.first->second << "," << range.second->second << "]");
+    
+    //  Copy MTDRecHits to the buffer array; select the seed hits
+    //  on the way, and store them in theSeeds.
+    for(auto itr = range.first; itr != range.second; ++itr) {
+      const unsigned hitidx = itr->second;
+      copy_to_buffer(begin+hitidx);
```

Comment:
valgrind points to invalid writes inside this call
and a bunch of invalid reads downstream in e.g. ```make_cluster``` calls